Sming/system/m_printf.cpp

```diff
@@ -117,143 +110,122 @@ int m_printf(const char* fmt, ...)
 
 int m_vsnprintf(char *buf, size_t maxLen, const char *fmt, va_list args)
 {
-	int i, base, flags;
-	char *str;
-	const char *s;
-	int8_t precision, width;
-	char pad;
-
-	char tempNum[40];
-
-	for (str = buf; *fmt; fmt++)
-	{
-		if(maxLen - (uint32_t)(str - buf) < OVERFLOW_GUARD)
-		{
-			*str++ = '(';
-			*str++ = '.';
-			*str++ = '.';
-			*str++ = '.';
-			*str++ = ')';
-
-			//mark end of string
-			*str = '\0';
-
-			//return maximum buffer len, so caller can detect not_enough_space
-			return maxLen;
-		}
-
-		if (*fmt != '%')
-		{
-			*str++ = *fmt;
-			continue;
-		}
-
-		flags = 0;
-		fmt++; // This skips first '%'
-
-		//reset attributes to defaults
-		precision = -1;
-		width = 0;
-		pad = ' ';
-		base = 10;
-        bool minus = 0;
-
-		do
-		{
-            if ('-' == *fmt) minus = 1, fmt++;
-            
-			//skip width and flags data - not supported
-			while ('+' == *fmt || '#' == *fmt || '*' == *fmt || 'l' == *fmt)
-				fmt++;
-
-			if (is_digit(*fmt)) {
-				if (*fmt == '0') {
-					pad = '0';
-					fmt++;
-				}
-				width = skip_atoi(&fmt);
-			}
-
-			if('.' == *fmt)
-			{
-				fmt++;
-				if (is_digit(*fmt))
-					precision = skip_atoi(&fmt);
-			}
-			else
-				break;
-		}while(1);
-
-		switch (*fmt)
-		{
-		case 'c':
-			*str++ = (unsigned char) va_arg(args, int);
-			continue;
-
-		case 's': {
-			s = va_arg(args, char *);
-
-			if (!s) s = "(null)";
-            size_t len = strlen(s);
-            len     = MIN( len,   precision );
-            len     = MIN( len,   maxLen - size_t(str - buf) - OVERFLOW_GUARD);
-            width   = MIN( width, maxLen - size_t(str - buf) - OVERFLOW_GUARD);
-
-            int padding = width - len;
-            while (!minus && padding-- > 0) *str++ = ' ';
-            while (len--) *str++ = *s++;
-            while (minus && padding-- > 0) *str++ = ' ';
-
-			continue;
-        }    
-
-		case 'p':
-			s = ultoa((unsigned long) va_arg(args, void *), tempNum, 16);
-			while (*s && (maxLen - (uint32_t)(str - buf) > OVERFLOW_GUARD))
-				*str++ = *s++;
-			continue;
-
-		case 'o':
-			base = 8;
-			break;
-
-		case 'x':
-		case 'X':
-			base = 16;
-			break;
-
-		case 'd':
-		case 'i':
-			flags |= SIGN;
-		case 'u':
-			break;
-
-		case 'f':
-
-			s = dtostrf_p(va_arg(args, double), width, precision, tempNum, pad);
-			while (*s && (maxLen - (uint32_t)(str - buf) > OVERFLOW_GUARD))
-				*str++ = *s++;
-			continue;
-
-		default:
-			if (*fmt != '%')
-				*str++ = '%';
-			if (*fmt)
-				*str++ = *fmt;
-			else
-				--fmt;
-			continue;
-		}
-
-		if (flags & SIGN)
-			s = ltoa_wp(va_arg(args, int), tempNum, base, width, pad);
-		else
-			s = ultoa_wp(va_arg(args, unsigned int), tempNum, base, width, pad);
-
-		while (*s && (maxLen - (uint32_t)(str - buf) > OVERFLOW_GUARD))
-			*str++ = *s++;
-	}
-
-	*str = '\0';
-	return str - buf;
+    size_t size = 0;
+    auto add = [&](char c) {
+        if (++size < maxLen) *buf++ = c;
+    };
+
+    while ( char f = pgm_read_byte(fmt) ) {
```

Comment:
You a right, this will make every call to *printf a tiny bit slower. The code size is almost identical. I see the following advantages:

- it will reduce total flash size and complexity as a many additional methods are no longer needed
- all classes and methods using _m_printf_ will benefit immediately
- life will be much easier for users as they do not need to care which method to call and which format specifier to use
- format strings should go to flashmem anyway to reduce ram size, it is faster than the existing implementation for format strings in flashmem
- printf is mostly used for serial output where execution speed does not matter that much
- for time critical operations direct c string manipulation should be used instead of the printf functions

IMHO the main advantage is that it eliminated a major pitfall for users working with progmem.